### ISSUE TITLE
ROL install failure

### ISSUE THREAD

**Issue by ibaned** [trilinos/Trilinos#2785]
@trilinos/rol

## Expectations

`make install` on Trilinos with ROL enabled should run without errors.

## Current Behavior

```
file INSTALL cannot find
  "/.../Trilinos/packages/rol/src/compatibility/ROL_TeuchosBatchManager.hpp"
```

## Motivation and Context

This is breaking all nightly builds of Albany.

**Comment by dridzal** [comment]
This should be fixed in #2790 ; @dpkouri just reviewed it.  However, note that the testing is still in progress, and we can't merge until that's finished.

**Comment by dridzal** [comment]
@ibaned @lxmota 
Merged.  Can you let us know if 'make install' works?

**Comment by lxmota** [comment]
@dridzal I just pulled from Trilinos develop. When I try `make install`, I get the error:

```
CMake Error at packages/rol/adapters/tpetra/src/cmake_install.cmake:49 (file):
  file INSTALL cannot find
  "/home/amota/LCM/Trilinos/packages/rol/src/compatibility/ROL_TeuchosBatchManager.hpp".

```

**Comment by dridzal** [comment]
That file is not in that directory anymore.  You may have to clean your build directory.  I’m not sure why CMake thinks it’s there ...

**Comment by lxmota** [comment]
Ok, I will try that.

**Comment by ibaned** [comment]
All our Trilinos builds from last night still failed with
```
  file INSTALL cannot find
  "/.../repos/Trilinos/packages/rol/src/compatibility/ROL_TeuchosBatchManager.hpp".
```

**Comment by lxmota** [comment]
Yes, I can confirm that the failure is the same as @ibaned reports, even if that file is not supposed to exist anymore.

**Comment by dridzal** [comment]
@bartlettroscoe 
Do you know how CMake can think that a file is still there when it isn't?  We don't include it explicitly, it's a ll done by *.hpp?  Is there a "special" clean command that people should use?

**Comment by lxmota** [comment]
I have tried compiling by completely removing my build directory and starting from scratch. I still get the error.

Also, the nightly Albany tests start clean by cloning both the Trilinos and Albany repositories and removing previous build and install directories. See [http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany](http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany). They all get the same error.

I looked at `packages/rol/cmake/BuildOptions.cmake`. There is this setting there:

```
# Compatibility Directory 
SET( DIR ${${PACKAGE_NAME}_SOURCE_DIR}/src/compatibility )
```
Maybe this variable is being used inadvertently to glob files? This is the only place where I could find the string `src/compatibility` in ROL.

Or as @dridzal says, there may be some caching by CMake going on.

**Comment by dridzal** [comment]
@gregvw do you have any input?  Why is that file included from that directory?


- - -

On May 22, 2018 at 7:45:10 PM MDT, Alejandro Mota <notifications@github.com> wrote:

I have tried compiling by completely removing my build directory and starting from scratch. I still get the error.

Also, the nightly Albany tests start clean by cloning both the Trilinos and Albany repositories and removing previous build and install directories. See http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany. They all get the same error.

I looked at packages/rol/cmake/BuildOptions.cmake. There is this setting there:

# Compatibility Directory
SET( DIR ${${PACKAGE_NAME}_SOURCE_DIR}/src/compatibility )


Maybe this variable is being used inadvertently to glob files? This is the only place where I could find the string src/compatibility in ROL.

Or as @dridzal<https://github.com/dridzal> says, there may be some caching by CMake going on.

**Comment by ikalash** [comment]
Any updates on getting this fixed?  All the Albany nightlies are broken due to this issue.

**Comment by bartlettroscoe** [comment]
Here is the problem:

```
$ find packages/rol -name "CMakeLists.txt" -exec grep -nH ROL_TeuchosBatchManager {} \;
packages/rol/adapters/tpetra/src/CMakeLists.txt:45:  APPEND_SET(HEADERS ${DIR}/ROL_TeuchosBatchManager.hpp)  
packages/rol/adapters/teuchos/src/CMakeLists.txt:45:    sol/ROL_TeuchosBatchManager.hpp
```

Glob instead of explicitly list files and this will not happen again.

**Comment by gahansen** [comment]
Here are a couple of links to the red dashboards. You can drill down to see the install error in context from there.

http://cdash.sandia.gov/CDash-2-3-0/index.php?project=Albany
https://my.cdash.org/index.php?project=Albany

These start from clean builds - shouldn't be any caching issues at play with these.

**Comment by dridzal** [comment]
Thanks @gregvw , let me know when this is in.
@bartlettroscoe , is there a way to run the entire test suite against an install build?  If this is possible, we would love to do it in our ROL nightly testing.

**Comment by gregvw** [comment]
@bartlettroscoe - Thanks for the suggestion. Somehow adapters never used this approach, but I have updated them to follow the same pattern as in rol/src/CMakeLists.txt
@dridzal - I just did a full build and install on a cee machine and it looks like all of the adapter headers are in the ${TRILINOS_INSTALL}/include path. Pushed changes to ROL-Trilinos if you want to synch.

**Comment by dridzal** [comment]
Will sync now.

**Comment by dridzal** [comment]
What does `ctest -S` do?

**Comment by bartlettroscoe** [comment]
@dridzal said:

> What does ctest -S do? Assuming the install target.

That is how you run ctest to drive builds and tests to submit to CDash.  You are using this in your ROL builds that submit to CDash if you inspect your files.  For example:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/ctest/drivers/sems_ci/single_ci_iter_aao_crf450.sh#L20

So to add installation testing, we would add another (optional) call to `CTEST_BUILD()` passing in the `install` target.  Then we could retrofit any automated build of any TriBITS project to add testing that at least the `install` target works.  And if you use globing of header files, it is very rare that there is a problem.

**Comment by dridzal** [comment]
@lxmota @ikalash @ibaned @gahansen  We believe that https://github.com/trilinos/Trilinos/pull/2810 fixed this issue.  Can you confirm and close?

**Comment by bartlettroscoe** [comment]
Are Albany's automated builds testing directly against the Trilinos 'develop' branch?  Does Albany not have a stable version of Trilinos that automated testing and most developers and users use?

**Comment by ibaned** [comment]
Yes, they point at `develop`. Although we've considered setting up our own "master" as other apps have done, I think this is preferable because we don't have to spend resources maintaining said master and our CI iterations with Trilinos are much more direct.

**Comment by bartlettroscoe** [comment]
> Yes, they point at `develop`. Although we've considered setting up our own "master" as other apps have done, I think this is preferable because we don't have to spend resources maintaining said master and our CI iterations with Trilinos are much more direct.

@ibaned, but then it is not fair to run Trilinos developers through a fire drill when there are Albany failures due to this (lack of an) integration model.  All Albany needs is to have its own Trilinos fork and that would allow backing out updates of the main 'develop' version once this occurred.  I will send you, Andy, and Glen an email offline about how easy it would be to set this up and the advantages of doing so.

**Comment by agsalin** [comment]
@bartlettroscoe : I've always known that Albany is doing a service to Trilinos by being a front line of testing for installation and API changes, since Trilinos tests are not adequate. If we protected Albany developers with a blessed Trilinos, as you suggest, then this bug would still have existed and not have been fixed until some application forced the issue. If it wasn't for Albany calling this fire drill, what would have happened? EMPIRE, Sierra, Nalu, Sparc, Sierra, Albany would have all just used old stable developments until -- someone took the plunge and called a fire drill a week or two later, when the bug identification would have been that much harder. That isn't better.

**Comment by bartlettroscoe** [comment]
@agsalin, you can test integration with your APP without subjecting all of your APP developers to errors that are found or creating a fire-drill for Trilinos developers.

**Comment by dridzal** [comment]
@agsalin @bartlettroscoe -- the point is that you can have it _both ways_ with a minimal investment.  Call the fire drill ---totally justified and well timed in this case--- and maintain productivity on the application side.  This has been part of the motivation for having a separate ROL development repo, because a while back it seemed like our (admittedly complex and app-like) PDE-OPT tests were catching errors in the new solver stack weekly.  The workflow proposed by @bartlettroscoe allows for some breathing room ---it is not supposed to encourage complacency.
